src/object_store/src/object/mod.rs

```diff
@@ -72,101 +91,199 @@ impl BlockLocation {
+            pub async fn upload(&self, path: &str, obj: Bytes) -> ObjectResult<()> {
+                $macro!(self, upload, path, obj)
+            }
+
+            pub async fn read(
+                &self,
+                path: &str,
+                block_loc: Option<BlockLocation>,
+            ) -> ObjectResult<Bytes> {
+                $macro!(self, read, path, block_loc)
+            }
+
+            pub async fn readv(
+                &self,
+                path: &str,
+                block_locs: &[BlockLocation],
+            ) -> ObjectResult<Vec<Bytes>> {
+                $macro!(self, readv, path, block_locs)
+            }
+
+            pub async fn metadata(&self, path: &str) -> ObjectResult<ObjectMetadata> {
+                $macro!(self, metadata, path)
+            }
+
+            pub async fn delete(&self, path: &str) -> ObjectResult<()> {
+                $macro!(self, delete, path)
+            }
+        }
+    };
+}
 
-    pub fn local(&self) -> &Arc<dyn ObjectStore> {
-        &self.local
-    }
+pub type ObjectStoreRef = Arc<ObjectStoreImpl>;
 
-    pub fn remote(&self) -> &Arc<dyn ObjectStore> {
-        &self.remote
-    }
+pub enum LocalObjectStore {
+    InMem(MonitoredObjectStore<InMemObjectStore>),
+    Disk(MonitoredObjectStore<DiskObjectStore>),
 }
 
-#[async_trait::async_trait]
-impl ObjectStore for HybridObjectStore {
-    async fn upload(&self, path: &str, obj: Bytes) -> ObjectResult<()> {
-        if is_local_path(path) {
-            self.local.upload(path, obj).await
-        } else {
-            self.remote.upload(path, obj).await
+macro_rules! local_object_store_method_body {
+    ($object_store:expr, $method_name:ident, $path:expr $(, $args:expr)*) => {
+        {
+            let orig_path = $path;
+            let path = parse_object_store_path($path);
+            assert!(path.is_local(), "local object store only but the path is not local: {}", orig_path);
+            match $object_store {
+                LocalObjectStore::InMem(in_mem) => in_mem.$method_name(path.as_str() $(, $args)*).await,
+                LocalObjectStore::Disk(disk) => disk.$method_name(path.as_str() $(, $args)*).await,
+            }
         }
-    }
+    };
+}
 
-    async fn read(&self, path: &str, block_loc: Option<BlockLocation>) -> ObjectResult<Bytes> {
-        if is_local_path(path) {
-            self.local.read(path, block_loc).await
-        } else {
-            self.remote.read(path, block_loc).await
-        }
-    }
+object_store_impl_body!(LocalObjectStore, local_object_store_method_body);
 
-    async fn readv(&self, path: &str, block_locs: &[BlockLocation]) -> ObjectResult<Vec<Bytes>> {
-        if is_local_path(path) {
-            self.local.readv(path, block_locs).await
-        } else {
-            self.remote.readv(path, block_locs).await
-        }
-    }
+pub enum RemoteObjectStore {
+    InMem(MonitoredObjectStore<InMemObjectStore>),
+    Disk(MonitoredObjectStore<DiskObjectStore>),
```

Comment:
Why can we have InMem and Disk object store as remote object store?

Reply:
InMem is used as remote object store in unit test and e2e test. Disk might be used as remote object store when we want to deploy a simple system without S3 or MinIO, but want to use more storage space.